### ISSUE TITLE
Will Symlinks ever be supported?

### ISSUE THREAD

**Issue by ruffle-b** [kahing/goofys#178]
As the subject says.

Will Goofys ever support symlinks?

**Comment by kahing** [comment]
The problem is that when you do ls, goofys has to be able to tell the kernel if a file is a symlink, and all ListObject tells us is the file name, size, and etag. We can possibly stuff a known payload (and hence known etag) and treat all objects with that etag as symlink, although if any other files happen to have the same etag then we are kind of screwed.

What do you plan to use symlinks for and can you emulate this on the application side?

**Comment by ruffle-b** [comment]
Yes I see that it has to be some kind of magic in a tag saying that the file content is the target path for a symlink. If the tag is suitably obscure (Goofys-Symlink-Target-Path=/example/file) then I'd have thought the chances of it colliding with a 'genuine' tag is pretty remote.

FWIW, s3fs does support symlinks so it might be worthwhile having a quick look at how that does it.

My use case is migration of a single-server based system to AWS. The filesystem in /var/www has numerous 'client' trees that have symlinks to directories of php code. To host this filesystem on S3 without symlinks would mean rewriting loads of stuff :(

**Comment by kahing** [comment]
I meant etag not this new object tagging thing that aws s3 supports. I just double checked and I don't see that ListObjects return the tags? Otherwise you'd have to fallback to what s3fs does which is to send a HEAD request for each object, at that point you could just use good old metadata (and also kills your performance)

**Comment by AudriusButkevicius** [comment]
You could have a magical _$symlink$ suffix.
lstat calls just return the mode with symlink bit set (which you can devise based on whether the suffix exists or not)
reading the symlink target is a separate readlink syscall anyway at which point you can just lookup the content of the suffixed key (that holds the target as body of the key) or HEAD of the object if you store the target as a tag.

even `ls -lah` issues two separate syscalls:
```
lstat("symlink", {st_mode=S_IFLNK|0777, st_size=26, ...}) = 0
lgetxattr("symlink", "security.selinux", 0xb30120, 255) = -1 ENODATA (No data available)
readlink("symlink", "target", 27) = 26
```

**Comment by kahing** [comment]
The problem is slightly more complicated. When the filesystem looks up "symlink", we don't yet know if it's a symlink yet, so we will have to HEAD both "symlink" and "symlink_$symlink$" even though most of the time they would not be symlinks.

We can lookup "symlink" first and fallback to "symlink_$symlink$", like we do for "dir" and "dir/", but that's already messy and we will also need to hand "symlink" and "symlink_$symlink$" both existing. Also, the mode is returned in readdir as well.

Using a magic etag has its downside too, since we can't use SSE-C or SSE-KMS even if that's configured (otherwise the etag may be meaningless) and the IAM policy may not allow us to do that.

Anyways I am not totally opposed to implementing symlink, I just haven't seen good enough reason do that yet given the costs. For example, is it sufficient to set x-amz-website-redirect-location and make sure goofys works fine with that? I imagine the use case of this is read-only and most people don't actually care if the file shows up as a symlink, as long as it's reading data from the correct object?

**Comment by AudriusButkevicius** [comment]
So I don't think it matters if it shows up as symlink or not, as long as its not consuming extra space. My usecase is uploading tons of OS libraries that have this weird mylib.so.0.0.0.1 linking to mylib.so.0.0.1 linking to mylib.so.1 linking to mylib.so.

It's nice to have programs work from the s3 mount without having to dereference links and duplicate storage consumption. Perhaps symlinks could be optionally provided as the slow double head traversal for people who want that, as I personally would be ready to pay the price for that.

Other usecase is using symlinks for packages. I have postgres/latest symlink pointing at postgres/9.6.1 etc, so in the current state of affairs I end up uploading two versions. I know the usecases are wishy washy that could be worked around differently, but its sadly out of my control and I have to provide a solution that is equivalent to the legacy one.


Lastly, you could temporarily cache (seconds) existance of symlinks after the readdir call, so when someone stats(x) you know you saw x_$suffix$ not that long ago.

**Comment by kahing** [comment]
So if I read this correctly, as long as read is working correctly that's all you care about. Do you also need to be able to create these links within goofys? If not, then all goofys need to do is follow the redirect when we see `x-amz-website-redirect-location`, which seems like the right thing to do and cost nothing for people who are not using it.

goofys can set `x-amz-website-redirect-location` as well but it will cause the file to sometimes show up as symlink and sometimes not. Which maybe ok.

**Comment by kahing** [comment]
trying to use a twitter poll to see what other people think: https://twitter.com/s3goofys/status/883398280628928512

**Comment by AudriusButkevicius** [comment]
Well if you support reading, you should support writing too. I guess this could cover both hardlinks and symlinks.

**Comment by ghthor** [comment]
I'm experimenting with hacky way to support symlinks in my current project. I'm using goofys as a volume for a docker container. Within the container I'm using overlayfs with goofys as the lowerdir. I let my job execute and write/read out everything to an upperlayer that's backed by tmpfs. After the job is completed I rsync everything that isn't a link into the goofys fs, and rsync all the links into the docker container and commit them. It's kind of working, but it's a very very ugly hack just to prove a that goofys has the performance to support my use case, which is backing the .pants.d working directory running within a CI environment without any persistant volumes(except s3).

**Comment by felixbuenemann** [comment]
What is the current behavior for symlinks? Will the syscall fail or will it just duplicate the data?

**Comment by pierorex** [comment]
Any updates on this? We're very interested in having this functionality available :)
